sygnal/webpushpushkin.py

```diff
@@ -0,0 +1,215 @@
+            proxy_url_str=proxy_url,
+        )
+        self.http_agent_wrapper = HttpAgentWrapper(self.http_agent)
+
+        privkey_filename = self.get_config("vapid_private_key")
+        if not privkey_filename:
+            raise PushkinSetupException("'vapid_private_key' not set in config")
+        if not os.path.exists(privkey_filename):
+            raise PushkinSetupException("path in 'vapid_private_key' does not exist")
+        self.vapid_private_key = Vapid.from_file(private_key_file=privkey_filename)
+        vapid_contact_email = self.get_config("vapid_contact_email")
+        if not vapid_contact_email:
+            raise PushkinSetupException("'vapid_contact_email' not set in config")
+        self.vapid_claims = {"sub": "mailto:{}".format(vapid_contact_email)}
+
+    async def _dispatch_notification_unlimited(self, n, device, context):
+        p256dh = device.pushkey
+        endpoint = device.data["endpoint"]
+        auth = device.data["auth"]
+        subscription_info = {
+            "endpoint": endpoint,
+            "keys": {"p256dh": p256dh, "auth": auth},
+        }
+        payload = WebpushPushkin._build_payload(n, device)
+        data = json.dumps(payload)
+
+        # we use the semaphore to actually limit the number of concurrent
+        # requests, since the HTTPConnectionPool will actually just lead to more
+        # requests being created but not pooled – it does not perform limiting.
+        with QUEUE_TIME_HISTOGRAM.time():
+            with PENDING_REQUESTS_GAUGE.track_inprogress():
+                await self.connection_semaphore.acquire()
+
+        try:
+            with SEND_TIME_HISTOGRAM.time():
+                with ACTIVE_REQUESTS_GAUGE.track_inprogress():
+                    response_wrapper = webpush(
+                        subscription_info=subscription_info,
+                        data=data,
+                        vapid_private_key=self.vapid_private_key,
+                        vapid_claims=self.vapid_claims,
+                        requests_session=self.http_agent_wrapper,
+                    )
+                    response = await response_wrapper.deferred
+                    await readBody(response)
+        finally:
+            self.connection_semaphore.release()
+
+        failed_pushkeys = []
+        # assume 4xx is permanent and 5xx is temporary
+        if 400 <= response.code < 500:
+            failed_pushkeys.append(device.pushkey)
+        return failed_pushkeys
+
+    @staticmethod
+    def _build_payload(n, device):
+        """
+        Build the payload data to be sent.
+        Args:
+            n: Notification to build the payload for.
+            device (Device): Device information to which the constructed payload
+            will be sent.
+
+        Returns:
+            JSON-compatible dict
+        """
+        payload = {}
+
+        if device.data:
+            payload.update(device.data.get("default_payload", {}))
```

Comment:
Copied this over from GCM, but wondering what it would do if default_payload is set to something that is not a dictionary, like `null`?

Reply:
In both cases we should probably either assert the type or do `data.get("default_payload") or {}` (which means any false-y value will be replaced with `{}`). This would still break if `default_payload` was `"foo"` or whatever...

Reply:
Added better validate for the push data. Shall I change GCM to do the same in this PR as well or a different one?

Reply:
Oh I missed this. I think while we're here we should do it. 👍